src/utils/pathCheck.ts

```diff
@@ -0,0 +1,106 @@
+import fs from 'fs';
+import path from 'path';
+import childProcess from 'child_process';
+
+import semver from 'semver';
+import resources from '@/resources';
+
+const fsPromises = fs.promises;
+
+const flags: Record<string, string> = {
+  helm:    'version',
+  kim:     '-v',
+  kubectl: 'version',
+};
+const regexes: Record<string, RegExp> = {
+  // helm has to match both
+  // current: version.BuildInfo{Version:"v3.5.3", ...
+  // older:   Client: &version.Version{SemVer:"v2.16.12", ...
+  helm:    /Version.*:.*?"v(.+?)"/,
+  kim:     /version v(\S+)/,
+  kubectl: /Client Version.*?GitVersion:"v(.+?)"/,
+};
+
+export default async function shadowInfo(sourceDir: string, targetDir: string, binaryName: string): Promise<Array<string>> {
+  const notes: Array<string> = [];
+  // Don't have access to Electron.app in unit tests, so can't use the resources module
+  const referencePath = path.join(sourceDir, binaryName);
+
+  try {
+    await fsPromises.access(referencePath, fs.constants.R_OK | fs.constants.X_OK);
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      console.log(err);
+
+      return notes;
+    }
+  }
+  const proposedVersion = await getVersion(referencePath, binaryName);
+
+  if (!proposedVersion) {
+    return notes;
+  }
+  const paths: Array<string> = process.env.PATH?.split(path.delimiter) || [];
+
+  let sawCurrentDir = false;
+
+  for (const currentDir of paths) {
+    if (currentDir === targetDir) {
+      sawCurrentDir = true;
+      continue;
+    }
+    const currentPath = path.join(currentDir, binaryName);
+
+    try {
+      await fsPromises.access(currentPath, fs.constants.R_OK | fs.constants.X_OK);
+    } catch (err) {
+      continue;
+    }
+    const currentVersion = await getVersion(currentPath, binaryName);
+
+    // For kubectl, don't bother comparing versions, just existence is enough of a problem
+    // if it occurs earlier in the path, because our kubectl is actually a symlink to kuberlr
+    if (binaryName === 'kubectl') {
+      if (!sawCurrentDir) {
+        notes.push(`Existing instance of ${ binaryName } in ${ currentDir } has version ${ currentVersion }, shadows linked version ${ proposedVersion }.`);
+      }
+      continue;
+    }
+
+    if (!currentVersion) {
+      continue;
+    }
```

Comment:
Isn't this still a problem? If I have a `/usr/bin/helm` (before `/usr/local/bin`) that's just a symlink to `/bin/false`, things will still be bad even though it's not a valid version?

Reply:
Yup, I had thought of that and decided to draw the line at expecting the executable at that spot to be a version of the expected executable. Otherwise we're wandering into unbounded territory. So we only report from a small set of possible problems and leave the large set of unlikely problems for the user to deal with.